org.eclipse.xtext/src/org/eclipse/xtext/nodemodel/LookAheadInfo.java

```diff
@@ -0,0 +1,100 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Sebastian Zarnekow and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.xtext.nodemodel;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.eclipse.xtext.Action;
+
+import com.google.common.annotations.Beta;
+import com.google.common.base.Preconditions;
+
+/**
+ * Utility to work with the lookAhead information in the node model.
+ * 
+ * @author Sebastian Zarnekow - Initial contribution and API
+ * 
+ * @since 2.22
+ */
+@Beta
+public class LookAheadInfo {
+
+	/* TODO 
+	 * - add getLookAhead(INode node) with support for leaf-nodes
+	 * - add getLookBeyond(INode node)
+	 * - getParserInformation(int offset): ParserInformation with ParserInformation( ParserRule, TextRegion, Iterator**) 
+	 *   usable for content assist, e.g. it must support parameterized rule stack recreation
+	 * - getEntryPoint(int offset, int length, String newText) with support for partial parsing (consider lookBeyond)
+	 * - Iterator** returns an iterator over nodes from the entry point to the position such that
+	 *   the parser would reach the position as fast as possible, e.g. an iterator that skips over
+	 *   regions that are not relevant to the parsers decision
+	 * - INode.getLookAheadInfo(): LookAheadInfo as default method that uses the resource to obtain an 
+	 *   injectable LookAheadInfo.Factory from allow customization
+	 */
+	
+	public static class InconsistentLookAheadException extends RuntimeException {
```

Comment:
@szarnekow @cdietrich Should the inner class be marked with the `@since 2.22` as well, or it is enough if the outer class have the `@since 2.22` tag?

Reply:
It is enough for the enclosing type.